enterprise/functions/src/main/java/io/crate/window/NthValueFunctions.java

```diff
@@ -62,9 +65,18 @@ public Object execute(int idxInPartition,
                           WindowFrameState currentFrame,
                           List<? extends CollectExpression<Row, ?>> expressions,
                           Input... args) {
-        if (idxInPartition == 0 || currentFrame.upperBoundExclusive() > seenFrameUpperBound) {
+        boolean shrinkingWindow = isShrinkingWindow(currentFrame, seenFrameLowerBound, seenFrameUpperBound);
+        if (idxInPartition == 0 || currentFrame.upperBoundExclusive() > seenFrameUpperBound || shrinkingWindow) {
+            seenFrameLowerBound = currentFrame.lowerBound();
             seenFrameUpperBound = currentFrame.upperBoundExclusive();
-            Object[] nthRowCells = currentFrame.getRowAtIndexOrNull(frameIndexSupplier.apply(currentFrame, args));
+
+            int index = frameIndexSupplier.apply(currentFrame, args);
+            if (shrinkingWindow) {
+                // if the lower bound is increasing we need to account for it to get the position we're interested in
```

Comment:
Would be nice to also explain *why* we need to account for it. The comment as it is imho doesn't add any information that is not already there.